java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -914,14 +931,31 @@ public void visitMethodReference(MethodReferenceTree methodReferenceTree) {
       if("new".equals(methodReferenceTree.method().name())) {
         resolution = resolve.findMethod(methodEnv, getType(methodReferenceTree.expression()), "<init>", samMethodArgs);
         associateReference(methodReferenceTree.method(), resolution.symbol());
+        return;
       } else {
         resolveMethodSymbol(methodReferenceTree.method(), methodEnv, samMethodArgs, ImmutableList.<JavaType>of());
       }
+      JavaType methodRefType = (JavaType) methodRefTree.symbolType();
+      if (methodRefTree.symbolType().isUnknown() || methodRefType.isTagged(JavaType.DEFERRED)) {
+        return;
+      }
+      JavaSymbol.MethodJavaSymbol samMethod = getSamMethod(methodRefType);
+      if(samMethod == null) {
+        return;
+      }
+      JavaType samReturnType = (JavaType) samMethod.returnType().type();
+      JavaType capturedReturnType = resolve.resolveTypeSubstitution(samReturnType, methodRefType);
```

Comment:
`capturedReturnType` is only used inside the `if`, could this declaration be moved there?